node/common/src/service/archive.rs

```diff
@@ -0,0 +1,321 @@
+            .name("openmina_archive".to_owned())
+            .spawn(move || {
+                Self::run(archive_receiver, address);
+            })
+            .unwrap();
+
+        Self::new(archive_sender)
+    }
+}
+
+impl node::transition_frontier::archive::archive_service::ArchiveService for NodeService {
+    fn send_to_archive(&mut self, data: ArchiveTransitionFronntierDiff) {
+        if let Some(archive) = self.archive.as_mut() {
+            if let Err(e) = archive.archive_sender.send(data.clone()) {
+                node::core::warn!(
+                    node::core::log::system_time();
+                    summary = "Failed sending diff to archive service",
+                    error = e.to_string()
+                );
+            }
+        }
+    }
+}
+
+// We need to replicate the ocaml node's RPC like interface
+mod rpc {
+    use binprot::BinProtWrite;
+    use mina_p2p_messages::rpc_kernel::{Message, NeedsLength, Query, RpcMethod};
+    use mina_p2p_messages::v2::{self, ArchiveRpc};
+    use mio::event::Event;
+    use mio::net::TcpStream;
+    use mio::{Events, Interest, Poll, Registry, Token};
+    use std::io::{self, Read, Write};
+    use std::net::SocketAddr;
+
+    const HANDSHAKE_MSG: [u8; 15] = [7, 0, 0, 0, 0, 0, 0, 0, 2, 253, 82, 80, 67, 0, 1];
+
+    pub enum HandleResult {
+        MessageSent,
+        ConnectionClosed,
+        ConnectionAlive,
+    }
+
+    impl HandleResult {
+        pub fn should_retry(&self) -> bool {
+            matches!(self, Self::ConnectionClosed)
+        }
+    }
+
+    pub fn send_diff(address: SocketAddr, data: v2::ArchiveRpc) -> io::Result<HandleResult> {
+        let rpc = encode_to_rpc(data);
+        process_rpc(address, &rpc)
+    }
+
+    fn encode_to_rpc(data: ArchiveRpc) -> Vec<u8> {
+        type Method = mina_p2p_messages::rpc::SendArchiveDiffUnversioned;
+        let mut v = vec![0; 8];
+
+        if let Err(e) = Message::Query(Query {
+            tag: Method::NAME.into(),
+            version: Method::VERSION,
+            id: 1,
+            data: NeedsLength(data),
+        })
+        .binprot_write(&mut v)
+        {
+            node::core::warn!(
+                node::core::log::system_time();
+                summary = "Failed binprot serializastion",
+                error = e.to_string()
+            )
+        }
+
+        let payload_length = (v.len() - 8) as u64;
+        v[..8].copy_from_slice(&payload_length.to_le_bytes());
+        v.splice(0..0, [1, 0, 0, 0, 0, 0, 0, 0, 0].iter().cloned());
+
+        v
+    }
+
+    fn process_rpc(address: SocketAddr, data: &[u8]) -> io::Result<HandleResult> {
+        println!("[archive-service] Data length: {}", data.len());
+        let mut poll = Poll::new()?;
+        let mut events = Events::with_capacity(128);
+        let mut event_count = 0;
+
+        // We still need a token even for one connection
+        const TOKEN: Token = Token(0);
+
+        let mut stream = TcpStream::connect(address)?;
```

Comment:
do we really need to re-connect every time?

Reply:
@tizoc Yes, that was one reason. Also once synced, the node will send data only a few times every 3 minutes realistically. No need to keep the connection alive IMO.